### PR TITLE
[GH-53] Implement minimum example of Node IPC

### PR DESCRIPTION
**Pull request by HirotoShioi**

### CODE REVIEW

File: src/NodeIPC/Test.hs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import           NodeIPC.Lib
+import           NodeIPC.Message
+
+--------------------------------------------------------------------------------
+-- Testing
+--------------------------------------------------------------------------------
+
+testIPC :: IO MsgOut
+testIPC = do
+    setEnv "NODE_CHANNEL_FD" "1" -- ???
```
Comment (ksaric) [inline src/NodeIPC/Test.hs]: You can simply abstract over it. You can just say "you, the caller, choose how you want to pass the environment information to the function, and I will support it". That provides you with the ability to test this and not to think about the way the rest of the system is going to work.
Having said that, think about how you would test the environment functions?
Take a look at the `IOLayer` from our previous project, for example :wink:

File: src/NodeIPC/Example.hs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    pong <- readClientMessage -- Pong
+    return (started, pong)
+  where
+    getReadWriteHandles :: IO (ReadHandle, WriteHandle)
+    getReadWriteHandles = do
+        (readFd, writeFd) <- createPipeFd
+        writeHndl         <- fdToHandle writeFd
+        readHndl          <- fdToHandle readFd
+
+        mapM_ (\h -> hSetBuffering h LineBuffering)
```
Comment (ksaric) [inline src/NodeIPC/Example.hs]: Make it readable, not smart.

File: src/NodeIPC/Example.hs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+-- reading the response that other had sent.
+exampleWithFD :: IO (MsgOut, MsgOut)
+exampleWithFD = do
+
+    (clientReadHandle, clientWriteHandle) <- getReadWriteHandles
+    (serverReadHandle, serverWriteHandle) <- getReadWriteHandles
+
+    -- Start the server
+    let nodePort = Port 8090
+    startNodeJsIPC serverReadHandle clientWriteHandle nodePort
```
Comment (ksaric) [inline src/NodeIPC/Example.hs]: This starts in a new thread, down the line.

**Comment by HirotoShioi** [comment]
Since coveralls are failing, I guess I need to implement a test case on a example.

File: src/NodeIPC/Lib.hs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+data NodeIPCException
+    = NodeChannelNotFound
+    -- ^ Node channel was not found
+    | UnableToParseNodeChannel Text
+    -- ^ Unable to parse given 'Text' as File descriptor
+    | IPCException
+    -- ^ Exception thrown when there's something wrong with IPC
+
+instance Show NodeIPCException where
```
Comment (ksaric) [inline src/NodeIPC/Lib.hs]: Nice.

File: src/NodeIPC/Lib.hs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            , ParseError safeText
+            ]
+        where
+          genSafeText :: Gen Text
+          genSafeText = strConv Lenient <$> listOf1 arbitraryASCIIChar
+
+opts :: Options
+opts = defaultOptions { sumEncoding = ObjectWithSingleField }
+
+instance ToJSON   MsgIn  where toEncoding = genericToEncoding opts
```
Comment (ksaric) [inline src/NodeIPC/Lib.hs]: I would add the newline here as usual.

File: src/NodeIPC/Lib.hs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+startNodeJsIPC readHandle writeHandle port =
+    liftIO $ void $ ipcListener readHandle writeHandle port
+
+-- | Start IPC listener with given Handles and Port
+--
+-- when the listener recieves 'Ping' it will return 'Pong'.
+--
+-- If it recieves 'QueryPort', then the listener
+-- responds with 'ReplyPort' with 'Port',
+ipcListener :: forall m . (MonadIO m, MonadCatch m)
```
Comment (ksaric) [inline src/NodeIPC/Lib.hs]: You can format this like the rest of the functions for better readability.

File: src/NodeIPC/Message.hs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.ByteString.Lazy.Char8 as BSLC
+import           Distribution.System (OS (Windows), buildOS)
+import           System.IO (hFlush, hGetLine)
+
+import qualified Prelude as P (Show (..))
+
+-- | Exception
+data MessageException
+    = DecodeFail BSL.ByteString
```
Comment (ksaric) [inline src/NodeIPC/Message.hs]: Good, good...

File: src/NodeIPC/Lib.hs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+                send Pong
+                shutdown
+
+    send :: MsgOut -> m ()
+    send = sendMessage writeHndl
+
+    -- Huge catch all, fix this:
+    handler :: IOError -> m ()
+    handler err = do
+        liftIO $ when (isEOFError err) $ logError "its an eof"
```
Comment (ksaric) [inline src/NodeIPC/Lib.hs]: You can pull this liftIO before do and remove it from all these lines if you want.

**Pull request merged**